### PR TITLE
feat: npm workspace migration

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/KIT-1811 (more details in Epic)

PR build works:
Screen Shot 2022-07-05 at 9 48 06 AM

### CODE REVIEW

File: .github/actions/e2e-atomic-angular/action.yml [hunk 1]
```diff
@@ -8,13 +8,17 @@ inputs:
...
+      shell: bash
+      working-directory: ./packages/samples/angular
     - uses: cypress-io/github-action@v4
       name: Run Cypress
       with:
         working-directory: ./packages/samples/angular
         start: npm run dev
         wait-on: 'http://localhost:4200'
-        install-command: npm install
+        install: false
```
Comment (ThibodeauJF) [inline .github/actions/e2e-atomic-angular/action.yml:21]: cypress expects a lockfile besides the package.json, so I move the step before

File: tsconfig.json [hunk 2]
```diff
@@ -1,3 +1,6 @@
 {
-  "extends": "./node_modules/gts/tsconfig-google.json"
+  "extends": "./node_modules/gts/tsconfig-google.json",
+  "compilerOptions": {
+    "skipLibCheck": true,
```
Comment (ThibodeauJF) [inline tsconfig.json:4]: was in most packages so I move this config at the top and removed it elsewhere

File: packages/atomic/package.json [hunk 3]
```diff
@@ -86,24 +84,22 @@
     "@storybook/addon-essentials": "6.3.12",
     "@storybook/addon-viewport": "6.3.12",
     "@storybook/api": "6.3.12",
+    "@storybook/builder-webpack5": "^6.3.12",
```
Comment (olamothe) [inline packages/atomic/package.json:87]: Is this needed to unify react version ?

If so, have you been able to `npm run storybook` and verify that it works okay ?

The storybook integration is quite custom !

File: packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts [hunk 4]
```diff
@@ -2,7 +2,7 @@ import {
   AnalyticsConfiguration,
   SearchEngineConfiguration,
 } from '@coveo/headless';
-import {AnalyticsClientSendEventHook} from '@coveo/headless/node_modules/coveo.analytics';
+import {AnalyticsClientSendEventHook} from '../../../../../../node_modules/coveo.analytics';
```
Comment (olamothe) [inline packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts:5]: just `import {AnalyticsClientSendEventHook} from 'coveo.analytics';` should work ? 🤔

**Review by olamothe — APPROVED** [review]
Awesome job ! 

It's grunt work, but hopefully it will solve many build problems we've had past few weeks. Lerna is a pain...

File: packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts (hunk 4 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts:5]: "coveo.analytics" is extraneous, it's not a direct dependency, but through headless

File: packages/atomic/package.json (hunk 3 above)
Comment (ThibodeauJF) [reply packages/atomic/package.json:87]: To unify the webpack version. Storybook still builds and runs fine 😄

File: packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts (hunk 4 above)
Comment (louis-bompart) [reply packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts:5]: > "coveo.analytics" is extraneous,

I think either headless should export AnalyticsClientSendEventHook, or you should install coveo.analytics in this workspace.

Comment (olamothe) [reply packages/atomic/src/components/search/atomic-search-interface/analytics-config.ts:5]: Yeah should be a direct dependency.

We can `import type` to be explicit that it's just for types, not code.

OR : Re-export the type from headless (since it is a dependency from Headless)
That second option would make sense, since it's actually a type that you need to know about if you implement on top of Headless (as showcased here)

File: package.json [hunk 5]
```diff
@@ -20,39 +20,50 @@
...
     "prettier": "2.6.2",
     "rimraf": "3.0.2",
     "semver": "7.3.7",
+    "ts-jest": "27.1.4",
     "typescript": "4.6.4"
   },
+  "workspaces": [
+    "packages/*",
+    "packages/atomic-angular/*",
+    "packages/atomic-angular/projects/*",
```
Comment (louis-bompart) [inline package.json:64]: FYI, nested workspaces are not supported by NPM
https://github.com/npm/cli/issues/4518

**Review by louis-bompart — COMMENTED** [review]
LGTM, I would be very careful with nested packages. If it works currently nice, but draft a plan of action to change that soon: given that this hierarchy ain't supported by npm, this is likely to cause pain in the future

File: lerna.json [hunk 6]
```diff
@@ -1,9 +1,9 @@
 {
   "packages": [
     "packages/*",
-    "packages/samples/**",
     "packages/atomic-angular/*",
-    "packages/atomic-angular/projects/*"
+    "packages/atomic-angular/projects/*",
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply lerna.json:5]: Bumping versions

File: package.json (hunk 5 above)
Comment (olamothe) [reply package.json:64]: In this case, we can only put `"packages/atomic-angular/projects/*",` methink, don't need ` "packages/atomic-angular/*",`

File: packages/samples/atomic-next/package.json [hunk 7]
```diff
@@ -3,25 +3,23 @@
...
+    "@types/react-dom": "18.0.5",
     "ncp": "2.0.0",
     "typescript": "4.6.4"
   },
   "scripts": {
-    "build": "next build",
-    "dev": "npm run copy:assets && npm run link:atomic-react && next dev",
-    "link:atomic-react": "cd ../../atomic-react && npm link ../samples/atomic-next/node_modules/react/ && ncp ../atomic/dist ../atomic-react/node_modules/@coveo/atomic/dist",
-    "copy:assets": "ncp node_modules/@coveo/atomic-react/dist/assets public/assets && ncp node_modules/@coveo/atomic-react/dist/lang public/lang public/assets"
+    "build": "tsc --noEmit && tsc --module commonjs --noEmit",
```
Comment (liviarett) [inline packages/samples/atomic-next/package.json:20]: out of curiosity, why did you change this?

**Review by liviarett — APPROVED** [review]

**Review by louis-bompart — APPROVED** [review]
project tree structure seems sounds, lgtm 👍

File: packages/samples/atomic-next/package.json (hunk 7 above)
Comment (ThibodeauJF) [reply packages/samples/atomic-next/package.json:20]: I'll improve this one after (had issues with the react version here too), for now I just wanted to make sure it compiles at least, like the atomic-react sample
